### PR TITLE
feat: Service CRM

### PR DESCRIPTION
**Pull request by haiderk0987**

### CODE REVIEW

File: erpnext/patches.txt [hunk 1]
```diff
@@ -789,3 +789,5 @@ erpnext.patches.v12_0.set_is_payable_in_lcv
 erpnext.patches.v12_0.trim_tables_maintenance_schedule
 erpnext.patches.v12_0.set_advance_tax_account_in_table
 erpnext.patches.v12_0.fix_purchase_invoice_double_gl
+erpnext.patches.v12_0.set_serial_no_in_transaction_documents
```
Comment (SaiFi0102) [inline erpnext/patches.txt:792]: We also need a patch to move values from Selling Settings to CRM Settings

File: erpnext/crm/doctype/opportunity/opportunity.js [hunk 2]
```diff
@@ -170,12 +174,31 @@ erpnext.crm.Opportunity = frappe.ui.form.Controller.extend({
 		});
 	},
 
+	setup_fields: function() {
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.js]: I think better we move this into `set_dynamic_field_label` and change the function name to `update_dynamic_fields`

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 3]
```diff
@@ -268,6 +293,9 @@ def set_missing_values(source, target):
 				"opportunity_from": "quotation_to",
 				"opportunity_type": "order_type",
 				"name": "opportunity",
+				"applies_to_serial_no": "applies_to_serial_no",
+				"applies_to_vehicle": "applies_to_vehicle",
+				"sales_person": "service_advisor"
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.py]: Remove sales person mapping .. or map it into the sales team table instead using `add_sales_person_from_source`

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 4]
```diff
@@ -348,28 +376,36 @@ def set_missing_values(source, target):
...
-			if target.meta.has_field('color'):
-				target.color = d.vehicle_color
-			elif target.meta.has_field('color_1'):
-				target.color_1 = d.vehicle_color
-			return
+@frappe.whitelist()
+def make_appointment(source_name, target_doc=None):
+	def set_missing_values(source, target):
+		add_sales_person_from_source(source, target)
+		target.appointment_type = frappe.get_cached_value("Opportunity Type", source.opportunity_type, "default_appointment_type")
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.py]: Set only if default_appointment_type is set .. check condition first

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 5]
```diff
@@ -348,28 +376,36 @@ def set_missing_values(source, target):
...
-			target.opportunity_item = d.name
-			if target.meta.has_field('color'):
-				target.color = d.vehicle_color
-			elif target.meta.has_field('color_1'):
-				target.color_1 = d.vehicle_color
-			return
+@frappe.whitelist()
+def make_appointment(source_name, target_doc=None):
+	def set_missing_values(source, target):
+		add_sales_person_from_source(source, target)
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.py]: I don't think Appointment has Sales Team table

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 6]
```diff
@@ -348,28 +376,36 @@ def set_missing_values(source, target):
...
+		target.appointment_type = frappe.get_cached_value("Opportunity Type", source.opportunity_type, "default_appointment_type")
+
+		target.run_method("set_missing_values")
+		target.run_method("calculate_taxes_and_totals")
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.py]: Appointment doesn't have `calculate_taxes_and_totals`

File: erpnext/crm/doctype/opportunity/opportunity.py [hunk 7]
```diff
@@ -348,28 +376,36 @@ def set_missing_values(source, target):
...
+
+	target_doc = get_mapped_doc("Opportunity", source_name, {
+		"Opportunity": {
+			"doctype": "Appointment",
+			"field_map": {
+				"name": "opportunity",
+				"applies_to_vehicle": "applies_to_vehicle",
```
Comment (SaiFi0102) [inline erpnext/crm/doctype/opportunity/opportunity.py:402]: Also add `applies_to_serial_no` in `field_map`

File: erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.js [hunk 8]
```diff
@@ -31,17 +31,20 @@ erpnext.maintenance.MaintenanceSchedule = frappe.ui.form.Controller.extend({
...
+				callback: function (r) {
+					if (r.message) {
+						me.frm.set_value(r.message);
+					}
 				}
-			}
-		});
+			});
+		}
 	},
```
Comment (SaiFi0102) [inline erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.js:49]: Lets add a button for manual creation of Opportunity from Maintenance Schedule

File: erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py [hunk 9]
```diff
@@ -205,3 +205,45 @@ def get_maintenance_schedule_from_serial_no(serial_no):
...
+		opportunity_doc.maintenance_schedule_row = schedule.name
+
+		project_template = frappe.get_cached_doc('Project Template', schedule.project_template)
+		for d in project_template.applicable_items:
+			opportunity_doc.append("items", {
+				"item_code": d.applicable_item_code,
+				"qty": d.applicable_qty,
+			})
+
+		opportunity_doc.save(ignore_permissions=True)
```
Comment (SaiFi0102) [inline erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.py:254]: lets also add `opportunity_doc.flags.ignore_mandatory = True`

**Review by SaiFi0102 — CHANGES_REQUESTED** [review]

File: erpnext/crm/doctype/opportunity/opportunity.py (hunk 3 above)
Comment (haiderk0987) [reply erpnext/crm/doctype/opportunity/opportunity.py]: But we don't have sales_team table in quotation

File: erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.js (hunk 8 above)
Comment (haiderk0987) [reply erpnext/maintenance/doctype/maintenance_schedule/maintenance_schedule.js:49]: added to pending list

**Review by SaiFi0102 — APPROVED** [review]

**Pull request merged**